### PR TITLE
[LoRA_Accuracy] add test_align_lora_accuracy.py

### LINKED ISSUE #617: ALign LoRA accuracy with HF
Align the LoRA accuracy with HF.

### `test_01_single_prompt_align_last_layer_logits_with_hf_on_cpu`

**Prompts**: Prompt1(LoRA).
**Number of Layers**: All layers.
**Environment**: TPU-V6e-1.
**Device**: Both of them use CPU.
**Comparison Object**: The logits of prefill and first decode token.
**Tolerance**: 1e-4.
**Methods**:

```python3
        tolerance = 1e-4
        np.testing.assert_allclose(
            hf_prefill_logits, sgl_prefill_logits, atol=tolerance, rtol=tolerance
        )
        np.testing.assert_allclose(
            hf_decode_logits, sgl_decode_logits, atol=tolerance, rtol=tolerance
        )
```

image

The following content are copied from [PR610](https://github.com/sgl-project/sglang-jax/pull/610).

### `test_02_multi_prompts_align_last_layer_logits_with_hf_on_tpu`

**Prompts**: Prompt1(LoRA), Prompt2(None), Prompt3(LoRA). Send them concurrently.
**Number of layers**: All layers.
**Environment**: TPU-V6e-1.
**Device**: HF uses CPU, SGLangJax uses TPU. Note: If SGLangJax used CPU, the results would be dramatically different. I do two experiments.
    - E1: When I use TPU for SGLangJax, the logits of first decode token are more reasonable. See the following figure.
    - E2: In order to check the wrong batch results are not brought by LoRA feature, I use the prompts without LoRAs and annotate the LoRA codes in engine, the results are the same.
    - In short, I uses TPU for SGLangJax to test the multi prompts.
**Comparison Object**: The logits of decode. Note: In SGLangJax, the different request may be prepared into different batches, and this makes the comparison is more difficult. So I use first decode token's logits.
**Tolerance**: 2e-9.
**Methods**:

```python3
            tolerance = 9e-2
            np.testing.assert_allclose(
                hf_decode_logits, sgl_decode_logits, atol=tolerance, rtol=tolerance
            )
```

image

### PR DESCRIPTION
**Pull request by aolemila**
Resolves #617.

## Motivation

Ensure the accuracy in SGLangJax LoRA.

## Modifications

Add `test_align_lora_accuracy.py` in `run_suite.py`(`unit-test-tpu-v6e-1`). I add two tests:

### `test_01_single_prompt_align_last_layer_logits_with_hf_on_cpu`

**Prompts**: Prompt1(LoRA).
**Number of Layers**: All layers.
**Environment**: TPU-V6e-1.
**Device**: Both of them use CPU.
**Comparison Object**: The logits of prefill and first decode token.
**Tolerance**: 1e-4.
**Methods**:

```python3
        tolerance = 1e-4
        np.testing.assert_allclose(
            hf_prefill_logits, sgl_prefill_logits, atol=tolerance, rtol=tolerance
        )
        np.testing.assert_allclose(
            hf_decode_logits, sgl_decode_logits, atol=tolerance, rtol=tolerance
        )
```

image

### `test_02_multi_prompts_align_last_layer_logits_with_hf_on_tpu`

**Prompts**: Prompt1(LoRA), Prompt2(None), Prompt3(LoRA). Send them concurrently.
**Number of layers**: All layers.
**Environment**: TPU-V6e-1.
**Device**: HF uses CPU, SGLangJax uses TPU. Note: If SGLangJax used CPU, the results would be dramatically different. I do two experiments.
    - E1: When I use TPU for SGLangJax, the logits of first decode token are more reasonable. See the following figure.
    - E2: In order to check the wrong batch results are not brought by LoRA feature, I use the prompts without LoRAs and annotate the LoRA codes in engine, the results are the same.
    - In short, I uses TPU for SGLangJax to test the multi prompts.
**Comparison Object**: The logits of decode. Note: In SGLangJax, the different request may be prepared into different batches, and this makes the comparison is more difficult. So I use first decode token's logits.
**Tolerance**: 2e-9.
**Methods**:

```python3
            tolerance = 9e-2
            np.testing.assert_allclose(
                hf_decode_logits, sgl_decode_logits, atol=tolerance, rtol=tolerance
            )
```

image

## Accuracy Tests

`python3 test/srt/lora/test_align_lora_accuracy.py`

## Checklist

- [ ] Please use English, otherwise it will be closed.
- [ ] The purpose of the PR, or link existing issues this PR will resolve.
- [ ] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/srt/server_args.py [hunk 1]
```diff
@@ -1122,6 +1122,9 @@ def check_dynamic_lora_args():
 
                     self.lora_paths = normalized_lora_refs
 
+                    # add max_loras_per_batch, '0' is reserved for requests without LoRA
+                    self.max_loras_per_batch += 1
+
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/server_args.py]: Adding this in `check_dynamic_lora_args` feels a bit confusing to me; I think handling it in lora_memory_pool would be cleaner.

Comment (aolemila) [reply python/sgl_jax/srt/server_args.py]: > Adding this in `check_dynamic_lora_args` feels a bit confusing to me; I think handling it in lora_memory_pool would be cleaner.

Ok. So range will be (1, self.max_loras_per_batch).

**Review by JamesBrianD — APPROVED** [review]
LGTM
